pkg/provisioner/ironic/ironic.go

```diff
@@ -1,46 +1,318 @@
+		bmcCreds:              bmcCreds,
+		client:                client,
+		log:                   log.WithValues("host", host.Name),
+		instanceImageSource:   f.instanceImageSource,
+		instanceImageChecksum: f.instanceImageChecksum,
+	}
+	return p, nil
 }
 
 // Register the host with Ironic.
-func (p *ironicProvisioner) ensureExists(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error) {
-	reqLogger := log.WithValues("host", host.Name)
-	if host.Status.Provisioning.ID == "" {
-		host.Status.Provisioning.ID = "temporary-fake-id"
-		reqLogger.Info("setting provisioning id",
-			"provisioningID", host.Status.Provisioning.ID)
+func (p *ironicProvisioner) ensureExists() (dirty bool, err error) {
+	var ironicNode *nodes.Node
+
+	// Try to load the node by UUID
+	if p.status.ID != "" {
+		// Look for the node to see if it exists (maybe Ironic was
+		// restarted)
+		p.log.Info("looking for existing node by ID", "ID", p.status.ID)
+		ironicNode, err = nodes.Get(p.client, p.status.ID).Extract()
+		switch err.(type) {
+		case nil:
+			p.log.Info("found existing node by ID")
+		case gophercloud.ErrDefault404:
+			p.log.Info("did not find existing node by ID")
+		default:
+			return false, errors.Wrap(err,
+				fmt.Sprintf("failed to find node by ID %s", p.status.ID))
+		}
+	}
+
+	// Try to load the node by name
+	if ironicNode == nil {
+		p.log.Info("looking for existing node by name", "name", p.host.Name)
+		ironicNode, err = nodes.Get(p.client, p.host.Name).Extract()
+		switch err.(type) {
+		case nil:
+			p.log.Info("found existing node by name")
+			// Store the ID so other methods can assume it is set and
+			// so we can find the node using that value next time.
+			p.status.ID = ironicNode.UUID
+			dirty = true
+			p.log.Info("setting provisioning id", "ID", p.status.ID)
+		case gophercloud.ErrDefault404:
+			p.log.Info("did not find existing node by name")
+			ironicNode = nil
+		default:
+			return false, errors.Wrap(err,
+				fmt.Sprintf("failed to find node by name %s", p.host.Name))
+		}
+	}
+
+	// If we have not found a node yet, we need to create one
+	if ironicNode == nil {
+		p.log.Info("registering host in ironic")
+		ironicNode, err = nodes.Create(
+			p.client,
+			nodes.CreateOpts{
+				Driver:        "ipmi",
+				BootInterface: "pxe",
+				Name:          p.host.Name,
+				DriverInfo: map[string]interface{}{
+					// FIXME(dhellmann): The names of the parameters
+					// depend on the ironic driver. We are going to
+					// need to isolate the logic for building the
+					// DriverInfo dict somewhere.
+					"ipmi_port":     p.bmcAccess.Port(),
+					"ipmi_username": p.bmcCreds.Username,
+					"ipmi_password": p.bmcCreds.Password,
+					"ipmi_address":  p.bmcAccess.Hostname(),
+					// FIXME(dhellmann): The names of the images are tied
+					// to the version of ironic we are using and are
+					// likely to change.
+					//
+					// FIXME(dhellmann): We need to get our IP on the
+					// provisioning network from somewhere.
```

Comment:
Yeah my original thought is we could just put the full address in a config file and have that be generated, but I guess we don't have a way to template things ATM?.. maybe ignition? (I don't like ignition myself tho :))

Reply:
A config file might be fine, or we could change the startup script for the container to figure it out (or even the main function of the operator process).